prow/jira/jira.go

```diff
@@ -107,7 +110,7 @@ func NewClient(endpoint string, opts ...Option) (Client, error) {
 	}
 
 	jiraClient, err := jira.NewClient(retryingClient.StandardClient(), endpoint)
-	return &client{upstream: jiraClient, url: endpoint}, err
+	return &client{delegate: &delegate{upstream: jiraClient, url: endpoint}, logger: log}, err
```

Comment:
It would be nice to find a way to set this without having to implement it on every single method, because that is almost guaranteed to rot over time and be incomplete.

How about:
* We add an `upstreamConstructor func(t http.RoundTripper) *jiraClient`
* In `WithFields` etc we call it with a roundTripper that sets a `used` variable upon first use
* `Used` on the client uses said variable

Reply:
Nice idea, thanks. I'll rework this.

Reply:
@alvaroaleman After analysis I can't see this working with jira client extracted to the `delegate` struct. I want to have a common client (for all plugins), but a unique `used` field. Maybe I missed something?

Reply:
That is why I suggest the `upstreamConstructor`

Reply:
But that will create a new jira client every time called, right?

Reply:
Yes, it doesn't allow to update the RoundTripper

Reply:
@alvaroaleman got back to this PR to make discussed change, almost done but I see one potential problem:
https://github.com/kubernetes/test-infra/blob/34e7342a4b709f2b449f2119a049cb37865bad23/prow/jira/jira.go#L109
returns not only the client but also an error. That would require `ForPlugin` to return an error as well, and possibly also `NewAgent` in https://github.com/kubernetes/test-infra/blob/34e7342a4b709f2b449f2119a049cb37865bad23/prow/plugins/plugins.go#L204
which will propagate to hook...
I think it is a pretty serious change in logic, and before I continue I would like to have your opinion.